desc/objectives/_neoclassical.py

```diff
@@ -0,0 +1,250 @@
+
+        rho, alpha = np.atleast_1d(rho, alpha)
+        self._dim_f = rho.size
+        self._keys_1dr = [
+            "iota",
+            "iota_r",
+            "<|grad(rho)|>",
+            "min_tz |B|",
+            "max_tz |B|",
+            "R0",  # TODO: GitHub PR #1094
+        ]
+        self._constants = {
+            "quad_weights": 1,
+            "rho": rho,
+            "alpha": alpha,
+            "zeta": np.linspace(
+                0, 2 * np.pi * num_transit, knots_per_transit * num_transit
+            ),
+            "quad": chebgauss2(num_quad),
+        }
+        self._hyperparameters = {
+            "num_pitch": num_pitch,
+            "batch": batch,
+            "num_well": num_well,
+        }
+
+        super().__init__(
+            things=eq,
+            target=target,
+            bounds=bounds,
+            weight=weight,
+            normalize=normalize,
+            normalize_target=normalize_target,
+            loss_function=loss_function,
+            deriv_mode=deriv_mode,
+            name=name,
+        )
+
+    def build(self, use_jit=True, verbose=1):
+        """Build constant arrays.
+
+        Parameters
+        ----------
+        use_jit : bool, optional
+            Whether to just-in-time compile the objective and derivatives.
+        verbose : int, optional
+            Level of output.
+
+        """
+        eq = self.things[0]
+        self._grid_1dr = LinearGrid(
+            rho=self._constants["rho"], M=eq.M_grid, N=eq.N_grid, NFP=eq.NFP, sym=eq.sym
+        )
+        self._target, self._bounds = _parse_callable_target_bounds(
+            self._target, self._bounds, self._constants["rho"]
+        )
+
+        timer = Timer()
+        if verbose > 0:
+            print("Precomputing transforms")
+        timer.start("Precomputing transforms")
+
+        self._constants["transforms_1dr"] = get_transforms(
+            self._keys_1dr, eq, self._grid_1dr
+        )
+        self._constants["profiles"] = get_profiles(
+            self._keys_1dr + ["effective ripple"], eq, self._grid_1dr
+        )
+
+        timer.stop("Precomputing transforms")
+        if verbose > 1:
+            timer.disp("Precomputing transforms")
+
+        super().build(use_jit=use_jit, verbose=verbose)
+
+    def compute(self, params, constants=None):
+        """Compute the effective ripple.
+
+        Parameters
+        ----------
+        params : dict
+            Dictionary of equilibrium degrees of freedom, e.g.
+            ``Equilibrium.params_dict``
+        constants : dict
+            Dictionary of constant data, e.g. transforms, profiles etc.
+            Defaults to ``self.constants``.
+
+        Returns
+        -------
+        result : ndarray
+            Effective ripple as a function of the flux surface label.
+
+        """
+        if constants is None:
+            constants = self.constants
+        eq = self.things[0]
+        # TODO: compute all deps of effective ripple here
```

Comment:
we build transforms to compute deps etc on a field line following grid. Another option would be to compute everything on a fixed desc grid. do the coordinate mapping in objective.compute. Then have the registered computed function for effective ripple not require it's dependencies to be computed on the source grid, but instead require the source grid and the coordinate mapped grid thsemselves and interpolate to the coordinate mapped nodes to get data along a field line. That approach is better because we need many nodes and need cost of transforms to be linearithmic instead of quadratic. It also avoid filling `data` with 1000s of intermediate quantities. This is an improvement that `Bounce2D` makes.